distributed/client.py

```diff
@@ -3546,25 +3547,39 @@ def _get_task_stream(self, start=None, stop=None, count=None, plot=False,
             raise gen.Return(msgs)
 
     @gen.coroutine
-    def _register_worker_callbacks(self, setup=None):
-        responses = yield self.scheduler.register_worker_callbacks(setup=dumps(setup))
+    def _register_worker_callbacks(self, setup=None, name=None):
+        callbacks = {}
+        names = {}
+        # prepare the callbacks and their names
+        if setup is not None:
+            serialized = dumps(setup)
+            callbacks['setup'] = serialized
+            if name is None:
+                h = md5(serialized)
+                name = funcname(setup) + '-' + h.hexdigest()
```

Comment:
This achieves the hashing based on function "body" @mrocklin  mentioned.

Reply:
Speed is not a concern here since it is only intended to be used at initialization. 

The hash produces two names for lib1.setup and lib2.setup; even when a name is not explicitly given. I would preferred `__qualname__`, but it is only defined in Python 3.3+.

Reply:
I recommend using `dask.base.tokenize` generally for hashing objects.  We tend to centralize our logic there.